crates/sui-core/src/state_accumulator/mod.rs

```diff
@@ -0,0 +1,156 @@
+                        .into_iter()
+                        .map(|(obj_ref, _)| obj_ref.2)
+                        .collect::<Vec<ObjectDigest>>()
+                })
+                .collect::<Vec<ObjectDigest>>(),
+        );
+        acc.remove_all(
+            state
+                .effects
+                .iter()
+                .flat_map(|fx| {
+                    fx.deleted
+                        .clone()
+                        .into_iter()
+                        .map(|obj_ref| obj_ref.2)
+                        .collect::<Vec<ObjectDigest>>()
+                })
+                .collect::<Vec<ObjectDigest>>(),
+        );
+        // TODO almost certainly not currectly handling "mutated" effects.
+        acc.insert_all(
+            state
+                .effects
+                .iter()
+                .flat_map(|fx| {
+                    fx.mutated
+                        .clone()
+                        .into_iter()
+                        .map(|(obj_ref, _)| obj_ref.2)
+                        .collect::<Vec<ObjectDigest>>()
+                })
+                .collect::<Vec<ObjectDigest>>(),
+        );
+
+        epoch_store.insert_state_hash_for_checkpoint(&state.checkpoint_seq_num, &acc)?;
+
+        epoch_store
+            .checkpoint_state_notify_read
+            .notify(&state.checkpoint_seq_num, &acc);
+
+        Ok(acc)
+    }
+
+    /// Unions all checkpoint accumulators at the end of the epoch to generate the
+    /// root state hash and saves it. This function is guaranteed to be idempotent (despite the
+    /// underlying data structure not being) as long as it is not called in a multi-threaded
+    /// context. Can be called on non-consecutive epochs, e.g. to accumulate epoch 3 after
+    /// having last accumulated epoch 1.
+    pub fn accumulate_epoch(
+        &self,
+        epoch: &EpochId,
+        last_checkpoint_of_epoch: CheckpointSequenceNumber,
+        epoch_store: Arc<AuthorityPerEpochStore>,
+    ) -> Result<Accumulator, TypedStoreError> {
+        if let Some((_checkpoint, acc)) = self
+            .authority_store
+            .perpetual_tables
+            .root_state_hash_by_epoch
+            .get(epoch)?
+        {
+            return Ok(acc);
+        }
+
+        // Get the next checkpoint to accumulate (first checkpoint of the epoch)
+        // by adding 1 to the highest checkpoint of the previous epoch
+        let (epoch, (next_to_accumulate, mut root_state_hash)) = self
+            .authority_store
+            .perpetual_tables
+            .root_state_hash_by_epoch
+            .iter()
+            .skip_to_last()
+            .next()
+            .map(|(epoch, (highest, hash))| (epoch, (highest.saturating_add(1), hash)))
+            .unwrap_or((0, (0, Accumulator::default())));
+
+        for i in next_to_accumulate..=last_checkpoint_of_epoch {
+            let acc = epoch_store
+                .get_state_hash_for_checkpoint(&i)
```

Comment:
should we to a scan instead of repeated point lookups? i'm not 100% sure this is necessary but with 30k checkpoints per epoch (a plausible number) I think it could matter.

Also, a scan would be a bit more defensive which I think is a good idea: it allows you to assert that the starting and ending points are correct, i.e. the first entry in the table must `== next_to_accumulate` and the last must equal `last_checkpoint_of_epoch`. If either of those fail something must be very wrong, so it certainly merits a panic.

Reply:
Don't scans require that they were written to sequentially / in order (in this case in sequence number order)? That is not the case here, since we are writing the accumulator to the db inside of the execution task in checkpoint executor, so they may be written out of order.

But should be straightforward to fix by writing to the db in the execution processing (main) task in checkpoint executor instead.

Reply:
You can always use a scan (aka `.iter()`) - i would expect it to be the fastest option even if the data is written randomly. Remember that after a compaction, the data will be in-order.